chalk-ir/src/visit.rs

```diff
@@ -228,6 +228,15 @@ where
         }
     }
 
+    /// Invoked for each occurence of a `Binders`, before visiting them.
+    fn before_binders(&mut self, _kinds: &crate::VariableKinds<I>) {}
+    /// Invoked for each occurence of a `Canonical`, before visiting them.
+    fn before_canonical(&mut self, _kinds: &crate::CanonicalVarKinds<I>) {}
+    /// Invoked for each occurence of a `FnPointer`, before visiting them.
+    fn before_fn_pointer_substs(&mut self, _number: usize) {}
+    /// Invoked for each occurence of a type introducing binders, after visiting them.
+    fn after_any_binders(&mut self) {}
+
```

Comment:
I wonder if it's worth just adding a `visit_binders` function like in rustc? (Probably) Is it all helpful to have these split out? I don't know if rustc calls `visit_binder` for fn pointer substs or canonical items, but we should.

Unfortunately, I think that our `visit_binders` would have to look like `fn visit_binder<T: SuperVisit<I>>(&mut self, binders: crate::VariableKinds<I>, value: T) -> ControlFlow<Self::BreakTy>);`, but that's probably fine.

Reply:
Alternatively, we could just change this to only have `before_binders` and do the "into `VariableKinds` conversion" happen before calling it.

Reply:
I tried something like that first, the problem is that it makes the `Visitor` trait not object safe.

We could do the `before_binders`, but it'd mean doing the conversion in every visitor instead of just the one where it's necessary.

Reply:
Well, I mean, you only need to do the conversion in the `Visit` impls, right?